File: perl/sorting_algorithms.pl
#!/usr/bin/perl
use strict;

sub selection_sort {
	my @list = @{$_[0]};
	for (my $i = 0; $i < scalar @list; $i++) {
		my $min = $i;
		for (my $j = $i + 1; $j < scalar @list; $j++) {
			if ($list[$j] < $list[$min]) {
				$min = $j;
			}
		}
		@list[$i, $min] = @list[$min, $i];
	}
	return @list;
}

sub insertion_sort {
	my @list = @{$_[0]};
	for (my $i = 1; $i < scalar @list; $i++) {
		my $insert_into = 0;
		for (my $j = $i - 1; $j >= 0; $j--) {
			if ($list[$i] >= $list[$j]) {
				$insert_into = $j + 1;
				last;
			}
		}
		my $removed = splice(@list, $i, 1);
		splice(@list, $insert_into, 0, $removed);
	}
	return @list;
}

sub merge_sort {
	my @list = @{$_[0]};
	if (scalar @list < 2) {
		return @list;
	}
	my @left = @list;#@list[0..(scalar @list) / 2 - 1];
	my @right = splice(@left, (scalar @left) / 2);#@list[(scalar @list) / 2..(scalar @list)];
	@left = merge_sort([@left]);
	@right = merge_sort([@right]);
	my $i = 0;
	my $j = 0;
	my @sorted = ();
	while ($i < scalar @left && $j < scalar @right) {
		if ($left[$i] <= $right[$j]) {
			push(@sorted, $left[$i]);
			$i++;
		} else {
			push(@sorted, $right[$j]);
			$j++;
		}
	}
	for (; $i < scalar @left; $i++) {
		push(@sorted, $left[$i]);
	}
	for (; $j < scalar @right; $j++) {
		push(@sorted, $right[$j]);
	}
	return @sorted;
}

sub quick_sort {
	my @list = @{$_[0]};
	if (scalar @list < 2) {
		return @list;
	}
	my $pivot = $list[0];
	my @left = ();
	my @right = ();
	for (my $i = 1; $i < scalar @list; $i++) {
		if ($list[$i] <= $pivot) {
			push(@left, $list[$i]);
		} else {
			push(@right, $list[$i]);
		}
	}
	@left = quick_sort([@left]);
	@right = quick_sort([@right]);
	my @sorted = (@left, $pivot, @right);
	return @sorted;
}

my @arr = (1, 6, 1, 8, 0, 3, 3, 9, 8, 8, 7, 4, 9, 8, 9, 4, 8, 4, 8, 2, 0, 4, 5);

my @arr_selection_sorted = selection_sort([@arr]);
my @arr_insertion_sorted = insertion_sort([@arr]);
my @arr_merge_sorted = merge_sort([@arr]);
my @arr_quick_sorted = quick_sort([@arr]);

print("Selection sort: @arr_selection_sorted\n");
print("Insertion sort: @arr_insertion_sorted\n");
print("Merge sort:     @arr_merge_sorted\n");
print("Quick sort:     @arr_quick_sorted\n");

File: c/sorting_algorithms.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

#define NUM_DIGITS 16

void copy_array(int src[], int dest[], int len);

void swap(int arr[], int a, int b) {
	int tmp = arr[a];
	arr[a] = arr[b];
	arr[b] = tmp;
}

void selection_sort(int arr[], int len) {
	for (int i = 0; i < len; i++) {
		int min_index = i;
		for (int j = i + 1; j < len; j++) {
			if (arr[j] < arr[min_index]) {
				min_index = j;
			}
		}
		if (min_index != i) {
			swap(arr, min_index, i);
		}
	}
}

void insertion_sort(int arr[], int len) {
	for (int i = 1; i < len; i++) {
		int x = arr[i];
		int j = i;
		while (arr[j - 1] > x && j > 0) {
			arr[j] = arr[j - 1];
			j--;
		}
		arr[j] = x;
	}
}

void merge_sort(int arr[], int len) {
	if (len < 2) {
		return;
	}
	int mid = len / 2;
	int* left = (int*) malloc(mid * sizeof(int));
	int* right = (int*) malloc((len - mid) * sizeof(int));
	copy_array(arr, left, mid);
	copy_array(arr + mid, right, len - mid);
	merge_sort(left, mid);
	merge_sort(right, len - mid);
	int a = 0;
	int b = 0;
	while (a < mid && b < len - mid) {
		arr[a + b] = left[a] <= right[b] ? left[a++] : right[b++];
	}
	for (; a < mid; a++) {
		arr[a + b] = left[a];
	}
	for (; b < len - mid; b++) {
		arr[a + b] = right[b];
	}
	free(left);
	free(right);
}

void quick_sort(int arr[], int len) {
	if (len < 2) {
		return;
	}
	int left[len];
	int right[len];
	int left_count = 0;
	int right_count = 0;
	int mid = len / 2;
	int pivot = arr[mid];
	for (int i = 0; i < len; i++) {
		if (i == mid) {
			continue;
		}
		if (arr[i] <= pivot) {
			left[left_count++] = arr[i];
		} else {
			right[right_count++] = arr[i];
		}
	}
	quick_sort(left, left_count);
	quick_sort(right, right_count);
	copy_array(left, arr, left_count);
	arr[left_count] = pivot;
	copy_array(right, arr + left_count + 1, right_count);
}

void copy_array(int src[], int dest[], int len) {
	for (int i = 0; i < len; i++) {
		dest[i] = src[i];
	}
}

void display_array(int arr[], int len) {
	printf("%d", arr[0]);
	for (int i = 1; i < len; i++) {
		printf(", %d", arr[i]);
	}
}

int main() {
	double GOLDEN_RATIO = (1 + sqrt(5)) / 2;
	int DIGITS[NUM_DIGITS];
	for (int i = 0; i < NUM_DIGITS; i++) {
		DIGITS[i] = (int) fmod(GOLDEN_RATIO * pow(10, i), 10);
	}
	
	int selection_sorted[NUM_DIGITS];
	copy_array(DIGITS, selection_sorted, NUM_DIGITS);
	selection_sort(selection_sorted, NUM_DIGITS);
	
	int insertion_sorted[NUM_DIGITS];
	copy_array(DIGITS, insertion_sorted, NUM_DIGITS);
	insertion_sort(insertion_sorted, NUM_DIGITS);
	
	int merge_sorted[NUM_DIGITS];
	copy_array(DIGITS, merge_sorted, NUM_DIGITS);
	merge_sort(merge_sorted, NUM_DIGITS);
	
	int quick_sorted[NUM_DIGITS];
	copy_array(DIGITS, quick_sorted, NUM_DIGITS);
	quick_sort(quick_sorted, NUM_DIGITS);


	printf("Original:       "); display_array(DIGITS, NUM_DIGITS); printf("\n");
	printf("Selection sort: "); display_array(selection_sorted, NUM_DIGITS); printf("\n");
	printf("Insertion sort: "); display_array(insertion_sorted, NUM_DIGITS); printf("\n");
	printf("Merge sort:     "); display_array(merge_sorted, NUM_DIGITS); printf("\n");
	printf("Quick sort:     "); display_array(quick_sorted, NUM_DIGITS); printf("\n");
}